src/main/java/seedu/address/logic/commands/PayCommand.java

```diff
@@ -117,38 +117,45 @@ private CommandResult executePayByIndex(Model model) throws CommandException {
 
     private CommandResult executePayAll(Model model) throws CommandException {
         assert nonNull(model);
-        // Create a deep copy of the filtered list
+        // Create a deep copy of the filtered list for viewing
         List<Person> lastShownList = new ArrayList<>(model.getFilteredPersonList());
+        List<Person> personsToBePaidList = new ArrayList<>();
+        List<Person> paidPersonsList = new ArrayList<>();
         List<Person> personsSkippedList = new ArrayList<>();
-        Person firstPersonToBePaid = null;
 
         // If all employees in the list are already paid, throw a CommandException
-        for (Person personToPay: lastShownList) {
-            if (!personToPay.isPaid()) {
-                firstPersonToBePaid = personToPay;
-                break;
+        for (Person person: lastShownList) {
+            if (!person.isPaid()) {
+                personsToBePaidList.add(person);
+            } else {
+                personsSkippedList.add(person);
             }
         }
 
-        if (firstPersonToBePaid == null) {
+        if (personsToBePaidList.isEmpty()) {
             throw new CommandException(MESSAGE_NO_ONE_TO_BE_PAID);
         }
 
         // Pay all persons in the list only if they have not been paid
-        for (Person personToPay: lastShownList) {
-            if (!personToPay.isPaid()) {
-                Person paidPerson = createPaidPerson(personToPay);
-                model.setPerson(personToPay, paidPerson);
-                if (personToPay.isSamePerson(firstPersonToBePaid)) {
-                    firstPersonToBePaid = paidPerson;
-                }
-            } else {
-                personsSkippedList.add(personToPay);
-            }
+        for (Person personToPay: personsToBePaidList) {
+            Person paidPerson = createPaidPerson(personToPay);
+            // Add changed persons to separate list first instead of
+            // immediately replacing into the model
+            paidPersonsList.add(paidPerson);
+        }
+
+        assert personsToBePaidList.size() == paidPersonsList.size();
+
+        // After all payment is done successfully, replace the
+        // changed persons into the model
+        for (int i = 0; i < personsToBePaidList.size(); i++) {
+            Person personToPay = personsToBePaidList.get(i);
+            Person paidPerson = paidPersonsList.get(i);
+            model.setPerson(personToPay, paidPerson);
         }
 
         // View the first person in the list that has been paid
-        model.setViewingPerson(firstPersonToBePaid);
+        model.setViewingPerson(paidPersonsList.get(0));
 
```

Comment:
I think that having an empty list would not really be considered an UI bug, as it confirms that there is no one left unpaid in the list of employees. But maybe we can consult the other members of the group as well regarding this?

Reply:
Actually, after thinking about it, it might be more confusing to the user to suddenly see a full employee list pop up after pay all. It would probably be better to allow the user to decide what to do next in this case. (Either use `list` to get the full list or `find` to get a filtered list)